modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala

```diff
@@ -129,11 +126,22 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
       _ <- if (!canJoinCluster) NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit]
       else Applicative[F].unit
 
-      hasPeerId <- clusterStorage.hasPeerId(toPeer.id)
-      _ <- if (hasPeerId) PeerIdInUse(toPeer.id).raiseError[F, Unit] else Applicative[F].unit
+      _ <- validateHandshakeConditions(toPeer)
+    } yield ()
+
+  private def validateHandshakeConditions(toPeer: PeerToJoin): F[Unit] =
+    validateHandshakeConditions(toPeer.id, toPeer.ip, toPeer.p2pPort)
+
+  private def validateHandshakeConditions(req: RegistrationRequest): F[Unit] =
+    validateHandshakeConditions(req.id, req.ip, req.p2pPort)
+
+  private def validateHandshakeConditions(id: PeerId, ip: Host, p2pPort: Port): F[Unit] =
+    for {
+      hasPeerId <- clusterStorage.hasPeerId(id)
+      _ <- if (hasPeerId) PeerIdInUse(id).raiseError[F, Unit] else Applicative[F].unit
 
-      hasPeerHostPort <- clusterStorage.hasPeerHostPort(toPeer.ip, toPeer.p2pPort)
-      _ <- if (hasPeerHostPort) PeerHostPortInUse(toPeer.ip, toPeer.p2pPort).raiseError[F, Unit]
+      hasPeerHostPort <- clusterStorage.hasPeerHostPort(ip, p2pPort)
+      _ <- if (hasPeerHostPort) PeerHostPortInUse(ip, p2pPort).raiseError[F, Unit]
```

Comment:
Maybe you could use `.ifM` here given that neither `hasPeerId` nor `hasPeerHostPort` are used anywhere further.

Reply:
Done :+1: